### PR TITLE
Refactor tests to run via pytest

### LINKED ISSUE #626: Port semgrep test script to pytest
There's a shell script called `run-lint-tests.sh` currently which runs tests for:

- Checking if JSON output for certain scenarios is correct
- Checking if well-written YAML rulesets are parsed successfully
- Checking if badly-written YAML rulesets are rejected gracefully
- Checking if the semgrep-rules repo's rules run successfully

The shell script is super useful, but very hard to work with. Lots of bits are copy-pasted, and test failure output doesn't explain what's wrong very well. I'd much rather write pytest functions that use subprocess calls, to be able to use fixtures, parametrization, and pytest's error reporting UI.

The goal here is to have `make test` call only `pytest`.

### PR DESCRIPTION
**Pull request by underyx**
Closes https://github.com/returntocorp/semgrep/issues/626

## TODO
- [x] Needs lots of documentation work before merging. But I'd rather do that after the general design & direction is approved. Namely, I went with a pytestic design which uses pytest features that optimize for simple test code, rather than verbose, type-safe code.
- [x] Need to figure out what to do with mypy. I'd argue it's not worth it to type-check everything in these test files.
- [x] Split commits for easier review.
- [x] Create issues fo next steps of testing architecture

### CODE REVIEW

File: semgrep/Makefile [hunk 1]
```diff
@@ -3,7 +3,7 @@ all:
 clean:
 	rm -rf ./build/
 test:
-	./tests/run-lint-tests.sh
+	PYTHONPATH=. pytest tests/
```
Comment (underyx) [inline semgrep/Makefile:6]: Once semgrep is pip installable, I'd prefer to make sure everyone has it installed as editable in their environment instead of setting PYTHONPATH.

File: semgrep/tests/e2e/snapshots/test_json_output/test_default_rule__file/results.json [hunk 2]
```diff
@@ -0,0 +1 @@
+{"results": [{"check_id": "eqeq-is-bad", "path": "targets/basic/stupid.py", "start": {"line": 3, "col": 12}, "end": {"line": 3, "col": 26}, "extra": {"message": "useless comparison operation `a+b == a+b` or `a+b != a+b`; possible bug?", "metavars": {"$X": {"start": {"line": 3, "col": 12, "offset": 55}, "end": {"line": 3, "col": 17, "offset": 60}, "abstract_content": "a+b", "unique_id": {"type": "AST", "md5sum": "07d71d85769e594dba9d7ae3d295c01f"}}}, "metadata": {}, "severity": "ERROR"}}, {"check_id": "javascript-basic-eqeq-bad", "path": "targets/basic/stupid.js", "start": {"line": 3, "col": 13}, "end": {"line": 3, "col": 19}, "extra": {"message": "useless comparison", "metavars": {"$X": {"start": {"line": 3, "col": 13, "offset": 24}, "end": {"line": 3, "col": 14, "offset": 25}, "abstract_content": "x", "unique_id": {"type": "id", "value": "x", "kind": "Global", "sid": 1}}}, "metadata": {}, "severity": "ERROR"}}], "errors": []}
```
Comment (underyx) [inline semgrep/tests/e2e/snapshots/test_json_output/test_default_rule__file/results.json]: Note to self, find out why this printed message changed compared to the previous committed JSON.

File: semgrep/tests/e2e/snapshots/test_rule_parser/test_rule_parser__failure__error_messages/bad1/error.txt [hunk 3]
```diff
@@ -0,0 +1,26 @@
+Traceback (most recent call last):
```
Comment (underyx) [inline semgrep/tests/e2e/snapshots/test_rule_parser/test_rule_parser__failure__error_messages/bad1/error.txt:1]: Note that my paths being snapshotted here are okay, as this is marked as an expected failure and this issue is linked: https://github.com/returntocorp/semgrep/issues/678

File: semgrep/tests/e2e/test_semgrep_rules_repo.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+            "-m",
+            "semgrep",
+            "--dangerously-allow-arbitrary-code-execution-from-rules",
+            "--strict",
+            "--test",
+            "--test-ignore-todo",
+            "semgrep-rules",
+        ]
+    )
+    subprocess.check_output(["python", "-m", "semgrep", "--validate", "semgrep-rules"])
```
Comment (underyx) [inline semgrep/tests/e2e/test_semgrep_rules_repo.py]: I'm not fully sure what this does, @DrewDennison could you explain these commands and what they test?

**Comment by mschwager** [comment]
Overall this looks like a big improvement to me! Is there anyway we can break this down into multiple changesets so it's easier to review? Not only will that make it easier to review, but it'll also reduce the likelihood of us making a mistake. It seems like there are 3 pieces of work here:

1. Add necessary pytest functionality that the tests will build on
2. Refactor existing tests to use new pytest functionality
3. Move existing files to new location

Is it possible to do this in 3 separate changesets?

File: semgrep/requirements.txt [hunk 5]
```diff
@@ -1,5 +1,6 @@
 colorama==0.4.3
 pytest==5.3.5
+pytest-snapshot==0.4.0
```
Comment (mschwager) [inline semgrep/requirements.txt:3]: Is there any way we can achieve the same level of testing without snapshotting? If I'm understanding snapshots correctly, they seem very difficult to reason about. I can never tell why a huge JSON blob is no longer the same as a previous huge JSON blob. IMO we should be testing small amounts of functionality and return values instead of dumping everything to a file and essentially performing a `diff`.

File: semgrep/tests/e2e/conftest.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+def _run_semgrep(
+    config: Optional[Union[str, Path]] = None,
+    *,
+    target_name: str = "basic",
+    stderr: bool = False
+) -> str:
+    config_opts = ["--config", config] if config is not None else []
+
+    return subprocess.check_output(
```
Comment (mschwager) [inline semgrep/tests/e2e/conftest.py]: This is calling Python code from Python code. Is there any way we can just call this library directly instead of `subprocess`ing out?

**Review by mschwager — COMMENTED** [review]
Added a few inline comments.

Generally when I think about unit testing I try to avoid the filesystem as much as possible. I do this for a few reasons: 1.) hitting the filesystem is slow compared to in-memory operations, 2.) filesystems differ on different systems (e.g. Linux, BSD, macOS, Windows), which can lead to unreliable tests and/or lots of edge-cases, 3.) filesystems often require clean up, e.g. create files, delete files, what happens if there's an exception? what happens if the file already exists? 4.) using the filesystem can often lead to order-dependent tests, i.e. tests have to run in a certain order to pass. Generally all tests should run independent of each other and have no order-dependencies.

I'm not sure how feasible it is to avoid the filesystem altogether here, but in my experience that leads to much more stable testing. If necessary I'd pull in something like https://github.com/jmcgeheeiv/pyfakefs/. Thoughts?

File: semgrep/tests/e2e/snapshots/test_json_output/test_default_rule__file/results.json (hunk 2 above)
Comment (underyx) [reply semgrep/tests/e2e/snapshots/test_json_output/test_default_rule__file/results.json]: Ah of course! I changed this test to just test whether the default semgrep rules location is picked up correctly, and split the `--generate-config` test to a separate file. So this now uses eqeq.yml instead of the semgrep-rules template.

File: semgrep/requirements.txt (hunk 5 above)
Comment (nbrahms) [reply semgrep/requirements.txt:3]: This seems like the age-old debate about end-to-end vs. unit tests. IMO, in general you need both: end-to-end tests (using snapshots here) guarantee behavior, while unit tests allow you to debug regressions.

But end-to-end tests would usually be primary, as without them, you can't guarantee your system works. So we want those most urgently.

Comment (nbrahms) [reply semgrep/requirements.txt:3]: cf. https://blog.colinbreck.com/should-we-write-a-unit-test-or-an-end-to-end-test/

File: semgrep/tests/conftest.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+def _run_semgrep(
+    config: Optional[Union[str, Path]] = None,
+    *,
+    target_name: str = "basic",
+    stderr: bool = False
+) -> str:
+    config_opts = ["--config", config] if config is not None else []
+
+    return subprocess.check_output(
```
Comment (nbrahms) [inline semgrep/tests/conftest.py]: At some point in the future, we probably want to be able to assert non-0 return codes

File: semgrep/tests/qa/test_public_repos.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        xfail_repo(
+            "https://github.com/apache/incubator-superset",
+            reason=(
+                "https://github.com/returntocorp/semgrep/issues/581, "
+                "https://github.com/returntocorp/semgrep/issues/582"
+            ),
+        ),
+    ],
+)
+def test_semgrep_on_repo(run_semgrep, repo_url):
```
Comment (nbrahms) [inline semgrep/tests/qa/test_public_repos.py]: Can this run from a temporary path?

**Review by nbrahms — APPROVED** [review]

File: semgrep/requirements.txt (hunk 5 above)
Comment (underyx) [reply semgrep/requirements.txt:3]: Thanks! I had similar thoughts and had just documented that philosophy here: https://github.com/returntocorp/semgrep/blob/bence/pytest/semgrep/tests/README.md

File: semgrep/tests/qa/test_public_repos.py (hunk 8 above)
Comment (underyx) [reply semgrep/tests/qa/test_public_repos.py]: It actually does, see behavior of https://github.com/returntocorp/semgrep/blob/bence/pytest/semgrep/tests/README.md#run_semgrep — the fixture prepares a temporary workspace with rules and target files.

This was one of the bits I was worried might be too magical 😅

File: semgrep/tests/conftest.py (hunk 7 above)
Comment (underyx) [reply semgrep/tests/conftest.py]: This is already possible! Check [`test_rule_parser.py`](https://github.com/returntocorp/semgrep/blob/bence/pytest/semgrep/tests/e2e/test_rule_parser.py).

Sample snippet:

```python
    with pytest.raises(CalledProcessError) as excinfo:
        run_semgrep(f"rules/syntax/{filename}.yaml")
    assert excinfo.value.returncode != 0
```

File: semgrep/tests/qa/test_public_repos.py (hunk 8 above)
Comment (nbrahms) [reply semgrep/tests/qa/test_public_repos.py]: ah... maybe can we rename `run_semgrep` to `run_semgrep_in_tmp`?

File: semgrep/tests/conftest.py (hunk 7 above)
Comment (nbrahms) [reply semgrep/tests/conftest.py]: ah right, because we throw a `CalledProcessError`. cool cool.

File: semgrep/requirements.txt (hunk 5 above)
Comment (nbrahms) [reply semgrep/requirements.txt:3]: FWIW, pretty-printing the JSON blobs might make debugging e2e failures much easier. Food for thought.

File: semgrep/tests/qa/test_public_repos.py (hunk 8 above)
Comment (underyx) [reply semgrep/tests/qa/test_public_repos.py]: Yeah! Renamed.

File: semgrep/requirements.txt (hunk 5 above)
Comment (underyx) [reply semgrep/requirements.txt:3]: Yeah. I was considering that too but felt like it would be too unclean to mess with the stdout 🤔

Added this change now though!

File: semgrep/tests/e2e/conftest.py (hunk 6 above)
Comment (underyx) [reply semgrep/tests/e2e/conftest.py]: I think it fits with the end-to-end philosophy to test whether commands run as expected. In `tests/unit/` we definitely shouldn't have a single usage of `subprocess` though. Does that sound okay?

**Comment by underyx** [comment]
@mschwager looking at your replies with the lens of end-to-end testing being prioritized here and adding brand new unit tests being out of scope from this PR[0]:
 
> Generally when I think about unit testing I try to avoid the filesystem as much as possible. I do this for a few reasons: 1.) hitting the filesystem is slow compared to in-memory operations

We have only around 5-10 file system writes in the whole test suite, other than the pytest-managed temp directory creations. This is definitely <0.1% of testing time currently, so it feels like premature optimization to address it. BTW a cool thing to note is that on many Linux systems we'd be using tmpfs which is actually an in-memory file system.

> 2.) filesystems differ on different systems (e.g. Linux, BSD, macOS, Windows), which can lead to unreliable tests and/or lots of edge-cases

While this is correct, semgrep is a tool that operates on files, so I'd consider it a plus if the tests will help us uncover cross-platform issues. Temp directory management is done by `pytest` which is well-maintained.

> 3.) filesystems often require clean up, e.g. create files, delete files, what happens if there's an exception? what happens if the file already exists?
> 4.) using the filesystem can often lead to order-dependent tests, i.e. tests have to run in a certain order to pass. Generally all tests should run independent of each other and have no order-dependencies.

This is all handled for us upstream, each test gets a clean directory to run in: https://docs.pytest.org/en/latest/tmpdir.html 

[0]: Though thanks for bringing it up, added it as a TODO item to create an issue for this and other things to improve in tests further.

**Comment by underyx** [comment]
@nbrahms do you have any guidance for me as for mypy? Should I fix all these errors, or is ignoring some justifiable?

```
semgrep/tests/e2e/test_autofix.py:1: error: Function is missing a type annotation
semgrep/tests/e2e/test_equivalence.py:1: error: Function is missing a type annotation
semgrep/tests/qa/test_semgrep_rules_repo.py:4: error: No library stub file for module 'pytest'
semgrep/tests/qa/test_semgrep_rules_repo.py:10: error: Function is missing a type annotation
semgrep/tests/e2e/test_rule_parser.py:3: error: No library stub file for module 'pytest'
semgrep/tests/e2e/test_rule_parser.py:7: error: Function is missing a type annotation
semgrep/tests/e2e/test_rule_parser.py:7: error: Function is untyped after decorator transformation
semgrep/tests/e2e/test_rule_parser.py:14: error: Function is missing a type annotation
semgrep/tests/e2e/test_rule_parser.py:14: error: Function is untyped after decorator transformation
semgrep/tests/e2e/test_rule_parser.py:37: error: Function is missing a type annotation
semgrep/tests/e2e/test_rule_parser.py:37: error: Function is untyped after decorator transformation
semgrep/tests/e2e/test_json_output.py:4: error: No library stub file for module 'pytest'
semgrep/tests/e2e/test_json_output.py:7: error: Function is missing a type annotation
semgrep/tests/e2e/test_json_output.py:11: error: Function is missing a type annotation
semgrep/tests/e2e/test_json_output.py:17: error: Function is missing a type annotation
semgrep/tests/e2e/test_json_output.py:23: error: Function is missing a type annotation
semgrep/tests/e2e/test_json_output.py:27: error: Function is missing a type annotation
semgrep/tests/e2e/test_json_output.py:34: error: Function is missing a type annotation
semgrep/tests/e2e/test_json_output.py:43: error: Function is missing a type annotation
semgrep/tests/e2e/test_json_output.py:49: error: Function is missing a type annotation
semgrep/tests/e2e/test_json_output.py:54: error: Function is missing a type annotation
semgrep/tests/e2e/test_generate_config.py:5: error: Function is missing a type annotation
semgrep/tests/qa/test_public_repos.py:5: error: No library stub file for module 'pytest'
semgrep/tests/qa/test_public_repos.py:10: error: Function is missing a type annotation
semgrep/tests/qa/test_public_repos.py:36: error: Call to untyped function "xfail_repo" in typed context
semgrep/tests/qa/test_public_repos.py:37: error: Call to untyped function "xfail_repo" in typed context
semgrep/tests/qa/test_public_repos.py:41: error: Call to untyped function "xfail_repo" in typed context
semgrep/tests/qa/test_public_repos.py:42: error: Call to untyped function "xfail_repo" in typed context
semgrep/tests/qa/test_public_repos.py:43: error: Call to untyped function "xfail_repo" in typed context
semgrep/tests/qa/test_public_repos.py:44: error: Call to untyped function "xfail_repo" in typed context
semgrep/tests/qa/test_public_repos.py:53: error: Call to untyped function "xfail_repo" in typed context
semgrep/tests/qa/test_public_repos.py:62: error: Function is missing a type annotation
semgrep/tests/qa/test_public_repos.py:62: error: Function is untyped after decorator transformation
semgrep/tests/conftest.py:8: error: No library stub file for module 'pytest'
semgrep/tests/conftest.py:8: note: (Stub files are from https://github.com/python/typeshed)
semgrep/tests/conftest.py:43: error: Function is missing a type annotation
semgrep/tests/conftest.py:43: error: Function is untyped after decorator transformation
semgrep/tests/conftest.py:54: error: Function is missing a type annotation
semgrep/tests/conftest.py:64: error: Function is missing a type annotation
semgrep/tests/conftest.py:68: error: Function is missing a type annotation```

**Comment by nbrahms** [comment]
> @nbrahms do you have any guidance for me as for mypy? Should I fix all these errors, or is ignoring some justifiable?
> 
> ```
> semgrep/tests/e2e/test_autofix.py:1: error: Function is missing a type annotation
> ...
> ```

It looks like most of these are solved by jamming a `-> None:` on the end of your test defs.

**Comment by underyx** [comment]
> It looks like most of these are solved by jamming a `-> None:` on the end of your test defs.

Nah, then it would move on to asking for type defs for all the arguments. Which is something I could do but doesn't feel like the best use of time.

The second one is pytest lacking type stubs. Again, not sure if writing stubs is worth the time investment since these functions are so complex 😕

File: semgrep/requirements.txt (hunk 5 above)
Comment (mschwager) [reply semgrep/requirements.txt:3]: When I think about testing I tend to use unit tests to test a single component, integration tests to test the interaction of two (or a few) components, and e2e tests to test many components or the system as a whole. Since semgrep is really only one component, or two if you count the core separately, I would lean toward unit tests.

> ... end-to-end tests (using snapshots here) guarantee behavior, while unit tests allow you to debug regressions ... But end-to-end tests would usually be primary, as without them, you can't guarantee your system works.

I'm not sure I agree here. In my opinion unit tests can also guarantee behavior, and e2e tests can be used to debug regressions. I've seen highly stable systems that use nothing but unit tests, and very fragile systems that use all e2e tests, or a mix of all three.

Regardless, I won't harp on the issue too much since it seems the decision's been made to go with this approach :+1:

**Comment by mschwager** [comment]
> @mschwager looking at your replies with the lens of end-to-end testing being prioritized here and adding brand new unit tests being out of scope from this PR

Sounds good :+1:

File: semgrep/tests/e2e/test_semgrep_rules_repo.py (hunk 4 above)
Comment (nbrahms) [reply semgrep/tests/e2e/test_semgrep_rules_repo.py]: i'm pretty sure this test makes sure that the entire semgrep-rules repo has validating config files

**Comment by nbrahms** [comment]
> > It looks like most of these are solved by jamming a `-> None:` on the end of your test defs.
> 
> Nah, then it would move on to asking for type defs for all the arguments. Which is something I could do but doesn't feel like the best use of time.
> 
> The second one is pytest lacking type stubs. Again, not sure if writing stubs is worth the time investment since these functions are so complex 😕

You should be able to crib the stubs from Bento, I think

**Comment by underyx** [comment]
> You should be able to crib the stubs from Bento, I think

@nbrahms sadly bento barely uses pytest features, the most complex things don't have stubs there (`pytest.mark.*`, `pytest.param`, `pytest.raises`, etc.)

**Comment by underyx** [comment]
So what I did now is exclude tests/ from mypy. Let me know what errors you think are worth fixing and I'll fix those!

**Comment by nbrahms** [comment]
You should just be able to disable the mypy features for pytest... then you still get typing for intramodular calls.

**Comment by underyx** [comment]
> You should just be able to disable the mypy features for pytest... then you still get typing for intramodular calls.

And how would I handle the test arguments then?
```
def test_exclude_include(run_semgrep_in_tmp, snapshot, options):
```
should be changed to
```
def test_exclude_include(run_semgrep_in_tmp: Any, snapshot: Any, options: Any) -> None:
```
?

**Comment by underyx** [comment]
@mschwager I've split this into many smaller commits as per above, can you have another look?

**Comment by mschwager** [comment]
> @mschwager I've split this into many smaller commits as per above, can you have another look?

I should've been more specific - I meant separate PRs. But generally this LGTM :+1:

**Comment by underyx** [comment]
> I should've been more specific - I meant separate PRs. But generally this LGTM 👍

Ohh, sorry 😅 I'm used to reviewing changes on a per-commit basis, so I just assumed that's what you meant.

**Comment by underyx** [comment]
On the daily standup call we agreed with @nbrahms to just create #699 and exclude tests from mypy scans.

Also created #694, #695, #696 for more tests related tech debt.
